repo/repo.go

```diff
@@ -100,6 +107,55 @@ func ImportRepoRules(filename string, repoCache *RemoteCache) ([]*rule.Rule, err
 	return rules, nil
 }
 
+// MergeRules merges a list of generated repo rules with the already defined repo rules,
+// and then updates each rule's underlying file. If the generated rule matches an existing
+// one, then it inherits the file where the existing rule was defined. If the rule is new then
+// its file is set as the destFile parameter. A list of the updated files is returned.
+func MergeRules(genRules []*rule.Rule, existingRules map[*rule.File][]string, destFile *rule.File, kinds map[string]rule.KindInfo) []*rule.File {
+	sort.Stable(byRuleName(genRules))
+
+	rulesByFile := make(map[*rule.File][]*rule.Rule)
+	for file, repoNames := range existingRules {
+		sort.Strings(repoNames)
+		if file.Path == destFile.Path && file.MacroName() != "" && file.MacroName() == destFile.MacroName() {
+			file = destFile
+		}
+		k := 0
+		for i, j := 0, 0; j < len(genRules); {
+			if i == len(repoNames) || repoNames[i] > genRules[j].Name() {
+				genRules[k] = genRules[j]
+				j++
+				k++
+			} else if repoNames[i] == genRules[j].Name() {
+				rulesByFile[file] = append(rulesByFile[file], genRules[j])
+				i++
+				j++
+			} else {
+				i++
+			}
+		}
+		genRules = genRules[:k]
+	}
+	for _, r := range genRules {
+		rulesByFile[destFile] = append(rulesByFile[destFile], r)
+	}
+
+	updatedFiles := make(map[string]*rule.File)
+	for f, rules := range rulesByFile {
+		if uf, ok := updatedFiles[f.Path]; ok {
+			f.SyncMacroFile(uf)
+		}
+		merger.MergeFile(f, nil, rules, merger.PreResolve, kinds)
+		f.Sync()
+		updatedFiles[f.Path] = f
```

Comment:
Let's make sure not to return more than one file with the same path. If a file contains many macros, we should choose one main file, copy the other synced macros there, then return that.